File: Source/Client/Game/States/GameState.vb
Imports Core
Imports Microsoft.Xna.Framework
Imports Mirage.Sharp.Asfw.IO.Encryption

Public Class GameState
    Private Shared ReadOnly stateLock As New Object
    
    Public Sub New()
    End Sub
    
    Public Shared ResolutionHeight As Integer
    Public Shared ResolutionWidth As Integer

    Public Shared MapAnim As Byte

    ' Global dialogue index
    Public Shared diaHeader As String
    Public Shared diaBody As String
    Public Shared diaBody2 As String
    Public Shared diaIndex As Long
    Public Shared diaData1 As Long
    Public Shared diaData2 As Long
    Public Shared diaData3 As Long
    Public Shared diaData4 As Long
    Public Shared diaData5 As Long
    Public Shared diaDataString As String
    Public Shared diaStyle As Byte

    ' shop
    Public Shared shopSelectedSlot As Long
    Public Shared shopSelectedItem As Long
    Public Shared shopIsSelling As Boolean

    ' right click menu
    Public Shared PlayerMenuIndex As Long

    ' description
    Public Shared descType As Byte
    Public Shared descItem As Long
    Public Shared descLastType As Byte
    Public Shared descLastItem As Long
    Public Shared descText() As Type.TextStruct

    ' New char
    Public Shared newCharSprite As Long
    Public Shared newCharJob As Long
    Public Shared newCharGender As Long

    ' chars
    Public Shared CharName(MAX_CHARS) As String
    Public Shared CharSprite(MAX_CHARS) As Long
    Public Shared CharAccess(MAX_CHARS) As Long
    Public Shared CharJob(MAX_CHARS) As Long
    Public Shared CharNum As Byte

    ' elastic bars
    Public Shared BarWidth_NpcHP(MAX_MAP_NPCS) As Long
    Public Shared BarWidth_PlayerHP(MAX_PLAYERS) As Long
    Public Shared BarWidth_PlayerSP(MAX_PLAYERS) As Long
    Public Shared BarWidth_NpcHP_Max(MAX_MAP_NPCS) As Long
    Public Shared BarWidth_PlayerHP_Max(MAX_PLAYERS) As Long
    Public Shared BarWidth_PlayerSP_Max(MAX_PLAYERS) As Long
    Public Shared BarWidth_GuiHP As Long
    Public Shared BarWidth_GuiSP As Long
    Public Shared BarWidth_GuiEXP As Long
    Public Shared BarWidth_GuiHP_Max As Long
    Public Shared BarWidth_GuiSP_Max As Long
    Public Shared BarWidth_GuiEXP_Max As Long

    Public Shared CurrentEvents As Integer

    ' Directional blocking
    Public Shared DirArrowX(4) As Byte
    Public Shared DirArrowY(4) As Byte

    Public Shared UseFade As Boolean
    Public Shared FadeType As Integer
    Public Shared FadeAmount As Integer
    Public Shared FlashTimer As Integer

    ' Targetting
    Public Shared MyTarget As Integer
    Public Shared MyTargetType As Integer

    ' Chat bubble
    Public Shared ChatBubbleindex As Integer
    
    Public Shared chatShowLine As String

    ' chat
    Public Shared inSmallChat As Boolean
    Public Shared actChatHeight As Long
    Public Shared actChatWidth As Long
    Public Shared ChatButtonUp As Boolean
    Public Shared ChatButtonDown As Boolean
    Public Shared ChatScroll As Long
    Public Shared Chat_HighIndex As Long

    Public Shared EditorTileX As Integer
    Public Shared EditorTileY As Integer
    Public Shared EditorTileWidth As Integer
    Public Shared EditorTileHeight As Integer
    Public Shared EditorWarpMap As Integer
    Public Shared EditorWarpX As Integer
    Public Shared EditorWarpY As Integer
    Public Shared EditorShop As Integer
    Public Shared EditorAnimation As Integer
    Public Shared EditorLight As Integer
    Public Shared EditorShadow As Byte
    Public Shared EditorFlicker As Byte
    Public Shared EditorAttribute As Byte
    Public Shared EditorTileSelStart As Microsoft.Xna.Framework.Point
    Public Shared EditorTileSelEnd As Microsoft.Xna.Framework.Point
    Public Shared CopyMap As Boolean
    Public Shared TmpMaxX As Byte
    Public Shared TmpMaxY As Byte
    Public Shared TileHistoryHighIndex As Integer

    ' Player variables
    Public Shared MyIndex As Integer ' Index of actual player

    Public Shared InBank As Boolean
    
    Public Shared SkillBuffer As Integer
    Public Shared SkillBufferTimer As Integer
    Public Shared StunDuration As Integer
    Public Shared NextlevelExp As Integer

    ' Stops movement when updating a map
    Public Shared CanMoveNow As Boolean

    ' Controls main gameloop
    Public Shared InGame As Boolean
    Public Shared InMenu As Boolean
    
    Public Shared MapData As Boolean
    Public Shared PlayerData As Boolean

    ' Draw map name location
    Public Shared DrawLocX As Single = 10
    Public Shared DrawLocY As Single = 0
    Public Shared DrawMapNameColor As Color

    ' Game direction vars
    Public Shared DirUp As Boolean
    Public Shared DirDown As Boolean
    Public Shared DirLeft As Boolean
    Public Shared DirRight As Boolean

    ' Used to freeze controls when getting a new map
    Public Shared GettingMap As Boolean

    ' Used to check if FPS needs to be drawn
    Public Shared Bfps As Boolean
    Public Shared Blps As Boolean
    Public Shared BLoc As Boolean

    ' FPS and Time-based movement vars
    Public Shared ElapsedTime As Integer

    ' Mouse cursor tile location
    Public Shared CurX As Integer
    Public Shared CurY As Integer
    Public Shared CurMouseX As Integer
    Public Shared CurMouseY As Integer

    ' Game editors
    Public Shared MyEditorType As Integer
    Public Shared EditorIndex As Integer

    ' Spawn
    Public Shared SpawnNpcNum As Integer
    Public Shared SpawnNpcDir As Integer

    ' Items
    Public Shared ItemEditorNum As Integer
    Public Shared ItemEditorValue As Integer

    ' Resources
    Public Shared ResourceEditorNum As Integer

    ' Used for map editor heal & trap & slide tiles
    Public Shared MapEditorHealType As Integer
    Public Shared MapEditorHealAmount As Integer
    Public Shared MapEditorSlideDir As Integer

    Public Shared Camera As Rectangle
    Public Shared TileView As RectStruct

    ' Pinging
    Public Shared PingStart As Integer
    Public Shared PingEnd As Integer
    Public Shared Ping As Integer

    ' Indexing
    Public Shared ActionMsgIndex As Byte
    Public Shared BloodIndex As Byte

    Public Shared TempMapData() As Byte

    Public Shared ShakeTimerEnabled As Boolean
    Public Shared ShakeTimer As Integer
    Public Shared ShakeCount As Byte
    Public Shared LastDir As Byte

    Public Shared ShowAnimLayers As Boolean
    Public Shared ShowAnimTimer As Integer

    Public Shared EKeyPair As New KeyPair()

    ' Stream Content
    Public Shared Item_Loaded(MAX_ITEMS)
    Public Shared NPC_Loaded(MAX_NPCS)
    Public Shared Resource_Loaded(MAX_RESOURCES)
    Public Shared Animation_Loaded(MAX_ANIMATIONS)
    Public Shared Skill_Loaded(MAX_SKILLS)
    Public Shared Shop_Loaded(MAX_SHOPS)
    Public Shared Pet_Loaded(MAX_PETS)
    Public Shared Moral_Loaded(MAX_MORALS)

    Public Shared AnimEditorFrame(1) As Integer
    Public Shared AnimEditorTimer(1) As Integer

    Public Shared MovementSpeed As Double
    Public Shared CurrentCameraX As Double
    Public Shared CurrentCameraY As Double

    ' Number of graphic files
    Public Shared NumTileSets As Integer
    Public Shared NumCharacters As Integer
    Public Shared NumPaperdolls As Integer
    Public Shared NumItems As Integer
    Public Shared NumResources As Integer
    Public Shared NumAnimations As Integer
    Public Shared NumSkills As Integer
    Public Shared NumFogs As Integer
    Public Shared NumEmotes As Integer
    Public Shared NumPanoramas As Integer
    Public Shared NumParallax As Integer
    Public Shared NumPictures As Integer
    Public Shared NumInterface As Integer
    Public Shared NumGradients As Integer
    Public Shared NumDesigns As Integer
    
    Public Shared GameDestroyed As Boolean

    Public Shared VbKeyRight As Boolean
    Public Shared VbKeyLeft As Boolean
    Public Shared VbKeyUp As Boolean
    Public Shared VbKeyDown As Boolean
    Public Shared VbKeyShift As Boolean
    Public Shared VbKeyControl As Boolean
    Public Shared VbKeyAlt As Boolean
    Public Shared VbKeyEnter As Boolean
    
    Public Shared LastLeftClickTime As Integer
    Public Const DoubleClickTImer As Integer = 500 ' Time in milliseconds for double-click detection

    Public Const ChatBubbleWidth As Integer = 300

    Public Const Chat_Timer As Long = 20000

    Public Const EffectTypeFadein As Integer = 1
    Public Const EffectTypeFadeout As Integer = 2
    Public Const EffectTypeFlash As Integer = 3
    Public Const EffectTypeFog As Integer = 4
    Public Const EffectTypeWeather As Integer = 5
    Public Const EffectTypeTint As Integer = 6

    ' Chat variables
    Public Const ChatWidth As Long = 316

    ' Bank constants
    Public Const BankTop As Long = 28
    Public Const BankLeft As Long = 9
    Public Const BankOffsetY As Long = 6
    Public Const BankOffsetX As Long = 6
    Public Const BankColumns As Long = 10

    ' Inventory constants
    Public Const InvTop As Long = 28
    Public Const InvLeft As Long = 9
    Public Const InvOffsetY As Long = 6
    Public Const InvOffsetX As Long = 6
    Public Const InvColumns As Long = 5

    ' Character consts
    Public Const EqTop As Long = 315
    Public Const EqLeft As Long = 11
    Public Const EqOffsetX As Long = 8
    Public Const EqColumns As Long = 4

    ' Skill constants
    Public Const SkillTop As Long = 28
    Public Const SkillLeft As Long = 9
    Public Const SkillOffsetY As Long = 6
    Public Const SkillOffsetX As Long = 6
    Public Const SkillColumns As Long = 5

    ' Hotbar constants
    Public Const HotbarTop As Long = 0
    Public Const HotbarLeft As Long = 8
    Public Const HotbarOffsetX As Long = 41

    ' Shop constants
    Public Const ShopTop As Long = 28
    Public Const ShopLeft As Long = 9
    Public Const ShopOffsetY As Long = 6
    Public Const ShopOffsetX As Long = 6
    Public Const ShopColumns As Long = 7

    ' Trade
    Public Const TradeTop As Long = 0
    Public Const TradeLeft As Long = 0
    Public Const TradeOffsetY As Long = 6
    Public Const TradeOffsetX As Long = 6
    Public Const TradeColumns As Long = 5

    ' Gfx Path and variables
    Public Const GfxExt As String = ".png"

    Public Shared MapGrid As Boolean
    Public Shared EyeDropper As Boolean
    Public Shared HistoryIndex As Integer
    Public Shared HideLayers As Boolean
    Public Shared Night As Boolean

    ' Speed moving vars
    Public Const WalkSpeed As Byte = 8
    Public Const RunSpeed As Byte = 16

    ' Tile size constants
    Public Const PicX As Integer = 32
    Public Const PicY As Integer = 32

    ' Sprite, item, skill size constants
    Public Const SizeX As Integer = 32
    Public Const SizeY As Integer = 32

    ' Map
    Public Const MaxTileHistory As Byte = 50
    Public Const TileSize As Byte = 32 ' Tile size is 32x32 pixels
    
    ' Autotiles
    Public Const AutoInner As Byte = 1

    Public Const AutoOuter As Byte = 2
    Public Const AutoHorizontal As Byte = 3
    Public Const AutoVertical As Byte = 4
    Public Const AutoFill As Byte = 5

    ' Autotile Type
    Public Const AutotileNone As Byte = 0

    Public Const AutotileNormal As Byte = 1
    Public Const AutotileFake As Byte = 2
    Public Const AutotileAnim As Byte = 3
    Public Const AutotileCliff As Byte = 4
    Public Const AutotileWaterfall As Byte = 5

    ' Rendering
    Public Const RenderStateNone As Integer = 0

    Public Const RenderStateNormal As Integer = 1
    Public Const RenderStateAutotile As Integer = 2

    ' Map animations
    Public Shared WaterfallFrame As Integer

    Public Shared AutoTileFrame As Integer
    
    Public Shared NumProjectiles As Integer
    Public Shared InitProjectileEditor As Boolean
    Public Const EditorProjectile As Byte = 10
    Public Shared ProjectileChanged(MAX_PROJECTILES) As Boolean
    
    Public Shared ResourceIndex As Integer
    Public Shared ResourcesInit As Boolean
    
    Public Const MaxWeatherParticles As Integer = 100
    Public Shared WeatherParticle(MaxWeatherParticles) As WeatherParticleStruct

    Public Shared  FogOffsetX As Integer
    Public Shared FogOffsetY As Integer

    Public Shared CurrentWeather As Integer
    Public Shared CurrentWeatherIntensity As Integer
    Public Shared CurrentFog As Integer
    Public Shared CurrentFogSpeed As Integer
    Public Shared CurrentFogOpacity As Integer
    Public Shared CurrentTintR As Integer
    Public Shared CurrentTintG As Integer
    Public Shared CurrentTintB As Integer
    Public Shared CurrentTintA As Integer
    Public Shared DrawThunder As Integer
    
    Public Shared InShop As Integer ' is the player in a shop?
    Public Shared ShopAction As Byte ' stores the current shop action

End Class

File: Source/Client/Game/States/GameState.cs
using Core;
using System.Reflection.Metadata;
using Color = Microsoft.Xna.Framework.Color;
using Point = Microsoft.Xna.Framework.Point;

namespace Client
{

    public class GameState
    {
        public GameState()
        {
        }

        public static int ResolutionHeight;
        public static int ResolutionWidth;

        public static bool MapAnim;

        // Global dialogue index
        public static string diaHeader;
        public static string diaBody;
        public static string diaBody2;
        public static long diaIndex;
        public static long diaData1;
        public static long diaData2;
        public static long diaData3;
        public static long diaData4;
        public static long diaData5;
        public static string diaDataString;
        public static byte diaStyle;

        // shop
        public static long shopSelectedSlot;
        public static long shopSelectedItem;
        public static bool shopIsSelling;

        // right click menu
        public static long PlayerMenuIndex;

        // description
        public static byte descType;
        public static long descItem;
        public static byte descLastType;
        public static long descLastItem;
        public static Core.Type.Text[] Description;

        // New char
        public static long NewCharSprite;
        public static long NewCharJob;
        public static long NewCnarGender;

        // chars
        public static string[] CharName = new string[(Core.Constant.MAX_CHARS)];
        public static long[] CharSprite = new long[(Core.Constant.MAX_CHARS)];
        public static long[] CharAccess = new long[(Core.Constant.MAX_CHARS)];
        public static long[] CharJob = new long[(Core.Constant.MAX_CHARS)];
        public static byte CharNum;

        // elastic bars
        public static long[] BarWidth_NpcHP = new long[(Core.Constant.MAX_MAP_NPCS)];
        public static long[] BarWidth_PlayerHP = new long[Core.Constant.MAX_PLAYERS];
        public static long[] BarWidth_PlayerSP = new long[Core.Constant.MAX_PLAYERS];
        public static long[] BarWidth_NpcHP_Max = new long[(Core.Constant.MAX_MAP_NPCS)];
        public static long[] BarWidth_PlayerHP_Max = new long[Core.Constant.MAX_PLAYERS];
        public static long[] BarWidth_PlayerSP_Max = new long[Core.Constant.MAX_PLAYERS];
        public static long BarWidth_GuiHP;
        public static long BarWidth_GuiSP;
        public static long BarWidth_GuiEXP;
        public static long BarWidth_GuiHP_Max;
        public static long BarWidth_GuiSP_Max;
        public static long BarWidth_GuiEXP_Max;

        public static int CurrentEvents;

        // Directional blocking
        public static byte[] DirArrowX = new byte[4];
        public static byte[] DirArrowY = new byte[4];

        public static bool UseFade;
        public static int FadeType;
        public static int FadeAmount;
        public static int FlashTimer;

        // Targetting
        public static int MyTarget;
        public static int MyTargetType;

        // Chat bubble
        public static int ChatBubbleindex;

        public static string ChatShowLine;

        public static string[] MapNames = new string[Core.Constant.MAX_MAPS];

        // chat
        public static bool inSmallChat;
        public static long actChatHeight;
        public static long actChatWidth;
        public static bool ChatButtonUp;
        public static bool ChatButtonDown;
        public static long ChatScroll;
        public static long Chat_HighIndex;

        public static int EditorTileX;
        public static int EditorTileY;
        public static int EditorTileWidth;
        public static int EditorTileHeight;
        public static int EditorWarpMap;
        public static int EditorWarpX;
        public static int EditorWarpY;
        public static int EditorShop;
        public static int EditorAnimation;
        public static byte EditorAttribute;
        public static Point EditorTileSelStart;
        public static Point EditorTileSelEnd;
        public static bool CopyMap;
        public static byte TmpMaxX;
        public static byte TmpMaxY;

        // Player variables
        public static int MyIndex; // Index of actual player

        public static bool InBank;

        public static int SkillBuffer;
        public static int SkillBufferTimer;
        public static int StunDuration;
        public static int NextlevelExp;

        // Stops movement when updating a map
        public static bool CanMoveNow;

        // Controls main gameloop
        public static bool InGame;
        public static bool InMenu;

        public static bool MapData;
        public static bool PlayerData;

        // Draw map name location
        public static float DrawLocX = 10f;
        public static float DrawLocY = 0f;
        public static Color DrawMapNameColor;

        // Game direction vars
        public static bool DirUp;
        public static bool DirDown;
        public static bool DirLeft;
        public static bool DirRight;

        // Used to freeze controls when getting a new map
        public static bool GettingMap;

        // Used to check if FPS needs to be drawn
        public static bool Bfps;
        public static bool Blps;
        public static bool BLoc;

        // FPS and Time-based movement vars
        public static int ElapsedTime;

        // Mouse cursor tile location
        public static int CurX;
        public static int CurY;
        public static int CurMouseX;
        public static int CurMouseY;
        public static bool Info;

        // Game editors
        public static int MyEditorType;
        public static int EditorIndex;
        public static bool AdminPanel;

        // Spawn
        public static int SpawnNpcNum;
        public static int SpawnNpcDir;

        // Items
        public static int ItemEditorNum;
        public static int ItemEditorValue;

        // Resources
        public static int ResourceEditorNum;

        // Used for map editor heal & trap & slide tiles
        public static int MapEditorHealType;
        public static int MapEditorHealAmount;
        public static int MapEditorSlideDir;

        public static Core.Type.Rect Camera;
        public static Core.Type.Rect TileView;

        // Pinging
        public static int PingStart;
        public static int PingEnd;
        public static int Ping;

        // Indexing
        public static byte ActionMsgIndex;
        public static byte BloodIndex;

        public static byte[] TempMapData;

        public static bool ShakeTimerEnabled;
        public static int ShakeTimer;
        public static byte ShakeCount;
        public static byte LastDir;

        public static bool ShowAnimLayers;
        public static int ShowAnimTimer;

        // Stream Content
        public static int[] Item_Loaded = new int[Core.Constant.MAX_ITEMS];
        public static int[] Npc_Loaded = new int[Core.Constant.MAX_NPCS];
        public static int[] Resource_Loaded = new int[Core.Constant.MAX_RESOURCES];
        public static int[] Animation_Loaded = new int[Core.Constant.MAX_RESOURCES];
        public static int[] Skill_Loaded = new int[Core.Constant.MAX_SKILLS];
        public static int[] Shop_Loaded = new int[Core.Constant.MAX_SHOPS];
        public static int[] Pet_Loaded = new int[Core.Constant.MAX_PETS];
        public static int[] Moral_Loaded = new int[(Core.Constant.MAX_MORALS)];
        public static int[] Projectile_Loaded = new int[(Core.Constant.MAX_PROJECTILES)];

        public static int[] AnimEditorFrame = new int[2];
        public static int[] AnimEditorTimer = new int[2];

        public static double CurrentCameraX;
        public static double CurrentCameraY;

        // Number of graphic files
        public static int NumTileSets;
        public static int NumCharacters;
        public static int NumPaperdolls;
        public static int NumItems;
        public static int NumResources;
        public static int NumAnimations;
        public static int NumSkills;
        public static int NumFogs;
        public static int NumEmotes;
        public static int NumPanoramas;
        public static int NumParallax;
        public static int NumPictures;
        public static int NumInterface;
        public static int NumGradients;
        public static int NumDesigns;

        public static bool VbKeyRight;
        public static bool VbKeyLeft;
        public static bool VbKeyUp;
        public static bool VbKeyDown;
        public static bool VbKeyShift;
        public static bool VbKeyControl;
        public static bool VbKeyAlt;
        public static bool VbKeyEnter;

        public static int LastLeftClickTime;
        public const int DoubleClickTImer = 500; // Time in milliseconds for double-click detection
        public static int ClickCount;

        public const int ChatBubbleWidth = 300;

        public const long Chat_Timer = 20000L;

        public const int EffectTypeFadein = 1;
        public const int EffectTypeFadeout = 2;
        public const int EffectTypeFlash = 3;
        public const int EffectTypeFog = 4;
        public const int EffectTypeWeather = 5;
        public const int EffectTypeTint = 6;

        // Bank constants
        public const long BankTop = 28L;
        public const long BankLeft = 9L;
        public const long BankOffsetY = 6L;
        public const long BankOffsetX = 6L;
        public const long BankColumns = 10L;

        // Inventory constants
        public const long InvTop = 28L;
        public const long InvLeft = 9L;
        public const long InvOffsetY = 6L;
        public const long InvOffsetX = 6L;
        public const long InvColumns = 5L;

        // Character consts
        public const long EqTop = 315L;
        public const long EqLeft = 10L;
        public const long EqOffsetX = 8L;
        public const long EqColumns = 4L;

        // Skill constants
        public const long SkillTop = 28L;
        public const long SkillLeft = 9L;
        public const long SkillOffsetY = 6L;
        public const long SkillOffsetX = 6L;
        public const long SkillColumns = 5L;

        // Hotbar constants
        public const long HotbarTop = 0L;
        public const long HotbarLeft = 8L;
        public const long HotbarOffsetX = 40L;

        // Shop constants
        public const long ShopTop = 28L;
        public const long ShopLeft = 9L;
        public const long ShopOffsetY = 6L;
        public const long ShopOffsetX = 6L;
        public const long ShopColumns = 7L;

        // Trade
        public const long TradeTop = 0L;
        public const long TradeLeft = 0L;
        public const long TradeOffsetY = 6L;
        public const long TradeOffsetX = 6L;
        public const long TradeColumns = 5L;

        // Gfx Path and variables
        public const string GfxExt = ".png";

        public static bool MapGrid;
        public static bool EyeDropper;
        public static int TileHistoryIndex;
        public static int TileHistoryHighIndex;
        public static bool HideLayers;

        // Tile size constants
        public const int PicX = 32;
        public const int PicY = 32;

        // Sprite, item, skill size constants
        public const int SizeX = 32;
        public const int SizeY = 32;

        // Map
        public const int MaxTileHistory = 500;
        public const byte TileSize = 32; // Tile size is 32x32 pixels

        // Autotiles
        public const byte AutoInner = 1;

        public const byte AutoOuter = 2;
        public const byte AutoHorizontal = 3;
        public const byte AutoVertical = 4;
        public const byte AutoFill = 5;

        // Autotile Type
        public const byte AutotileNone = 0;

        public const byte AutotileNormal = 1;
        public const byte AutotileFake = 2;
        public const byte AutotileAnim = 3;
        public const byte AutotileCliff = 4;
        public const byte AutotileWaterfall = 5;

        // Rendering
        public const int RenderStateNone = 0;

        public const int RenderStateNormal = 0;
        public const int RenderStateAutotile = 2;

        // Map animations
        public static int WaterfallFrame;

        public static int AutoTileFrame;

        public static int NumProjectiles;
        public static bool InitProjectileEditor;

        public static int ResourceIndex;
        public static bool ResourcesInit;

        public static Core.Type.WeatherParticle[] WeatherParticle = new Core.Type.WeatherParticle[Core.Constant.MAX_WEATHER_PARTICLES];

        public static int FogOffsetX;
        public static int FogOffsetY;

        public static int CurrentWeather;
        public static int CurrentWeatherIntensity;
        public static int CurrentFog;
        public static int CurrentFogSpeed;
        public static int CurrentFogOpacity;
        public static int CurrentTintR;
        public static int CurrentTintG;
        public static int CurrentTintB;
        public static int CurrentTintA;
        public static int DrawThunder;

        public static int InShop; // is the player in a shop?
        public static byte ShopAction; // stores the current shop action

        public static int MapEditorTab;
        public static int CurLayer;
        public static int CurAutotileType;
        public static int CurTileset;

        // Editors
        public static bool InitEditor;
        public static bool InitMapEditor;
        public static bool InitItemEditor;
        public static bool InitResourceEditor;
        public static bool InitNpcEditor;
        public static bool InitSkillEditor;
        public static bool InitShopEditor;
        public static bool InitAnimationEditor;
        public static bool InitJobEditor;
        public static bool InitMoralEditor;
        public static bool InitAdminForm;
        public static bool InitMapReport;
        public static bool InitEventEditor;
        public static bool InitScriptEditor;

        // Editor edited items array
        public static bool[] Item_Changed = new bool[Core.Constant.MAX_ITEMS];
        public static bool[] Npc_Changed = new bool[Core.Constant.MAX_NPCS];
        public static bool[] Resource_Changed = new bool[Core.Constant.MAX_RESOURCES];
        public static bool[] Animation_Changed = new bool[Core.Constant.MAX_ANIMATIONS];
        public static bool[] Skill_Changed = new bool[Core.Constant.MAX_SKILLS];
        public static bool[] Shop_Changed = new bool[Core.Constant.MAX_SHOPS];
        public static bool[] Pet_Changed = new bool[Core.Constant.MAX_PETS];
        public static bool[] Job_Changed = new bool[(Core.Constant.MAX_JOBS)];
        public static bool[] Moral_Changed = new bool[(Core.Constant.MAX_MORALS)];
        public static bool[] ProjectileChanged = new bool[Core.Constant.MAX_PROJECTILES];
    }
}